### PR TITLE
Run all end-to-end tests with ABIEncoderV2

### LINKED ISSUE #4700: Run all end-to-end tests with ABIEncoderV2 too

### PR DESCRIPTION
**Pull request by axic**
Closes #4700.

### CODE REVIEW

**Comment by axic** [comment]
It seems not all the tests can be compiled with it:
```
:2:1: Warning: Experimental features are turned on. Do not use experimental features on live deployments.
pragma experimental ABIEncoderV2;
^-------------------------------^
:76:2: Warning: This declaration shadows an existing declaration.
	function Registrar() public {
 ^ (Relevant source part starts here and spans across multiple lines).
:11:1: The shadowed declaration is here:
contract Registrar is NameRegister {
^ (Relevant source part starts here and spans across multiple lines).

:12:16: Error: Indexed reference types cannot yet be used with ABIEncoderV2.
	event Changed(string indexed name);
	              ^-----------------^
:13:23: Error: Indexed reference types cannot yet be used with ABIEncoderV2.
	event PrimaryChanged(string indexed name, address indexed addr);
	                     ^-----------------^
:24:21: Error: Indexed reference types cannot yet be used with ABIEncoderV2.
	event AuctionEnded(string indexed _name, address _winner);
	                   ^------------------^
:25:15: Error: Indexed reference types cannot yet be used with ABIEncoderV2.
	event NewBid(string indexed _name, address _bidder, uint _value);
	             ^------------------^
```

Is that an actual bug we have or the old encoder supports it?

Note, the tests don't care about those events, so they may as well be broken.

Update: there is a single test, `event_indexed_string`, which does seem to validate it.

**Comment by axic** [comment]
So it seems all the failures are for cases where the two encoders explicitly differ.

Good news is that every other case works 🎉 

Now need to figure out a way how to properly have this merged.

**Comment by axic** [comment]
@chriseth @ekpyron do you think this can be finished easily (i.e. certain tests have to have the ability to say they are not ABIEncoderV2 compatible)?

If not, should we merge the `soltest` command line option so that locally it can be run, but not make it part of the CI?

Or should we just close this.

**Comment by chriseth** [comment]
I think this should be a feature of the new extracted end to end test format.

Another option would be to just implement indexed reference parameters for events for abi encoder v2.

**Comment by chriseth** [comment]
Rebased (and removed the "indexed" modifications).

**Comment by chriseth** [comment]
It looks like after https://github.com/ethereum/solidity/pull/6000 and https://github.com/ethereum/solidity/pull/6029 the only problem left is something to do with a cleanup function for mappings - and looking at the code, this might actually be already fixed.

**Comment by chriseth** [comment]
I was wrong, needed https://github.com/ethereum/solidity/pull/6030.

After that, this should turn green (apart from the few tests that are specifically written to highlight differences in the two encoders).

**Comment by chriseth** [comment]
Rebased on top of the fixes.

File: scripts/tests.sh [hunk 1]
```diff
@@ -191,6 +191,14 @@ do
   for vm in $EVM_VERSIONS
   do
     printTask "--> Running tests using "$optimize" --evm-version "$vm"..."
+
+    # Run constantinople tests with ABIEncoderV2 turned on
```
Comment (chriseth) [inline scripts/tests.sh]: We should also run constantinople without ABIEncoderV2.

File: test/libsolidity/GasCosts.cpp [hunk 2]
```diff
@@ -70,13 +70,17 @@ BOOST_AUTO_TEST_CASE(string_storage)
 
 	if (Options::get().evmVersion() <= EVMVersion::byzantium())
 		CHECK_GAS(134435, 130591, 100);
+	else if (Options::get().useABIEncoderV2)
```
Comment (axic) [inline test/libsolidity/GasCosts.cpp:74]: Why isn't this the first if?

File: test/libsolidity/GasMeter.cpp [hunk 3]
```diff
@@ -73,9 +73,14 @@ class GasMeterTestFramework: public SolidityExecutionFramework
 		// costs for transaction
 		gas += gasForTransaction(m_compiler.object(m_compiler.lastContractName()).bytecode, true);
 
-		BOOST_REQUIRE(!gas.isInfinite);
-		BOOST_CHECK_LE(m_gasUsed, gas.value);
-		BOOST_CHECK_LE(gas.value - _tolerance, m_gasUsed);
+		// Skip the tests when we force ABIEncoderV2. We should enable this again
```
Comment (axic) [inline test/libsolidity/GasMeter.cpp]: Can you make the "we should enable" part a next line and prepend it with `TODO:` or something similar?

File: test/libsolidity/GasMeter.cpp [hunk 4]
```diff
@@ -97,9 +102,14 @@ class GasMeterTestFramework: public SolidityExecutionFramework
 			*m_compiler.runtimeAssemblyItems(m_compiler.lastContractName()),
 			_sig
 		);
-		BOOST_REQUIRE(!gas.isInfinite);
-		BOOST_CHECK_LE(m_gasUsed, gas.value);
-		BOOST_CHECK_LE(gas.value - _tolerance, m_gasUsed);
+		// Skip the tests when we force ABIEncoderV2. We should enable this again
```
Comment (axic) [inline test/libsolidity/GasMeter.cpp]: Same here.

File: test/libsolidity/GasCosts.cpp (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/GasCosts.cpp:74]: Because it is only correct on constantinople.

Comment (chriseth) [reply test/libsolidity/GasCosts.cpp:74]: The test assumes that you do not run the v2 tests on pre-constantinople.

Comment (axic) [reply test/libsolidity/GasCosts.cpp:74]: Ok, can you add a comment in both places that this the reason so that others' wont move it around?

**Review by chriseth — APPROVED** [review]

**Pull request merged**